doom/config.org
#+title:Doom Emacs config
#+property: header-args:emacs-lisp :tangle yes :comments link
#+property: header-args :tangle no :results silent :eval no-export
#+html:<div align=center><p><img alt="Emacs Logo" width="150" height="150" src="emacs_logo.svg"></p></div>

This is my literate configuration for Doom Emacs. This file generates config.el, init.el and packages.el. Please run the following to generate these files when initially setting up.

#+begin_src shell :tangle no
emacs --batch --eval "(progn (require 'org) (setq org-confirm-babel-evaluate nil) (org-babel-tangle-file \"~/.config/doom/config.org\"))"
#+end_src

* Doom configuration
** Init
:PROPERTIES:
:header-args:emacs-lisp: :tangle no
:END:
A literate init.el. This provides modular configurability for Doom. Lets you enable and install supported packages and tweak their functionalities with flags.
#+name: init.el
#+attr_html: :collapsed t
#+begin_src emacs-lisp :tangle "init.el" :noweb no-export :comments no
;;; init.el -*- lexical-binding: t; -*-

;; This file controls what Doom modules are enabled and what order they load in.
;; Press 'K' on a module to view its documentation, and 'gd' to browse its directory.

(doom! :completion
       <<doom-completion>>

       :ui
       <<doom-ui>>

       :editor
       <<doom-editor>>

       :emacs
       <<doom-emacs>>

       :term
       <<doom-term>>

       :checkers
       <<doom-checkers>>

       :tools
       <<doom-tools>>

       :os
       <<doom-os>>

       :lang
       <<doom-lang>>

       :email
       <<doom-email>>

       :app
       <<doom-app>>

       :config
       <<doom-config>>
       )
#+end_src

*** Completion
#+name: doom-completion
#+begin_src emacs-lisp
company           ; the ultimate code completion backend
;;helm              ; the *other* search engine for love and life
;;ido               ; the other *other* search engine...
(ivy +icons)               ; a search engine for love and life
#+end_src

*** UI
#+name: doom-ui
#+begin_src emacs-lisp
deft              ; notational velocity for Emacs
doom              ; what makes DOOM look the way it does
doom-dashboard    ; a nifty splash screen for Emacs
doom-quit         ; DOOM quit-message prompts when you quit Emacs
;;(emoji +unicode)  ; 🙂
;;fill-column       ; a `fill-column' indicator
hl-todo           ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/REVIEW
hydra
;;indent-guides     ; highlighted indent columns
;;ligatures         ; ligatures and symbols to make your code pretty again
;;minimap           ; show a map of the code on the side
modeline          ; snazzy, Atom-inspired modeline, plus API
;;nav-flash         ; blink cursor line after big motions
;;neotree           ; a project drawer, like NERDTree for vim
ophints           ; highlight the region an operation acts on
(popup +defaults)   ; tame sudden yet inevitable temporary windows
;;tabs              ; a tab bar for Emacs
treemacs          ; a project drawer, like neotree but cooler
;;unicode           ; extended unicode support for various languages
vc-gutter         ; vcs diff in the fringe
vi-tilde-fringe   ; fringe tildes to mark beyond EOB
window-select     ; visually switch windows
workspaces        ; tab emulation, persistence & separate workspaces
;;zen               ; distraction-free coding or writing
#+end_src

*** Editor
#+name: doom-editor
#+begin_src emacs-lisp
(evil +everywhere); come to the dark side, we have cookies
file-templates    ; auto-snippets for empty files
fold              ; (nigh) universal code folding
format  ; automated prettiness
;;god               ; run Emacs commands without modifier keys
;;lispy             ; vim for lisp, for people who don't like vim
multiple-cursors  ; editing in many places at once
;;objed             ; text object editing for the innocent
;;parinfer          ; turn lisp into python, sort of
;;rotate-text       ; cycle region at point between text candidates
snippets          ; my elves. They type so I don't have to
word-wrap         ; soft wrapping with language-aware indent
#+end_src

*** Emacs
#+name: doom-emacs
#+begin_src emacs-lisp
(dired +icons)            ; making dired pretty [functional]
electric          ; smarter, keyword-based electric-indent
;;ibuffer         ; interactive buffer management
(undo +tree)              ; persistent, smarter undo for your inevitable mistakes
vc                ; version-control and Emacs, sitting in a tree
#+end_src

*** Term
#+name: doom-term
#+begin_src emacs-lisp
eshell            ; the elisp shell that works everywhere
;;shell             ; simple shell REPL for Emacs
;;term              ; basic terminal emulator for Emacs
vterm             ; the best terminal emulation in Emacs
#+end_src

*** Checkers
#+name: doom-checkers
#+begin_src emacs-lisp
syntax              ; tasing you for every semicolon you forget
;;(spell +flyspell) ; tasing you for misspelling mispelling
;;grammar           ; tasing grammar mistake every you make
#+end_src

*** Tools
#+name: doom-tools
#+begin_src emacs-lisp
;;ansible
;;debugger          ; FIXME stepping through code, to help you add bugs
;;direnv
;;docker
;;editorconfig      ; let someone else argue about tabs vs spaces
;;ein               ; tame Jupyter notebooks with emacs
(eval +overlay)     ; run code, run (also, repls)
;;gist              ; interacting with github gists
(lookup +dictionary)              ; navigate your code and its documentation
lsp
magit             ; a git porcelain for Emacs
make              ; run make tasks from Emacs
;;pass              ; password manager for nerds
pdf               ; pdf enhancements
;;prodigy           ; FIXME managing external services & code builders
;;rgb               ; creating color strings
;;taskrunner        ; taskrunner for all your projects
;;terraform         ; infrastructure as code
;;tmux              ; an API for interacting with tmux
;;upload            ; map local to remote projects via ssh/ftp
#+end_src

*** OS support
#+name: doom-os
#+begin_src emacs-lisp
(:if IS-MAC macos)  ; improve compatibility with macOS
;;tty               ; improve the terminal Emacs experience
#+end_src

*** Language support
#+name: doom-lang
#+begin_src emacs-lisp
;;agda              ; types of types of types of types...
(cc +lsp)                ; C/C++/Obj-C madness
;;clojure           ; java with a lisp
common-lisp       ; if you've seen one lisp, you've seen them all
;;coq               ; proofs-as-programs
;;crystal           ; ruby at the speed of c
;;csharp            ; unity, .NET, and mono shenanigans
;;data              ; config/data formats
;;(dart +flutter)   ; paint ui and not much else
;;elixir            ; erlang done right
;;elm               ; care for a cup of TEA?
emacs-lisp        ; drown in parentheses
;;erlang            ; an elegant language for a more civilized age
;;ess               ; emacs speaks statistics
;;faust             ; dsp, but you get to keep your soul
;;fsharp            ; ML stands for Microsoft's Language
;;fstar             ; (dependent) types and (monadic) effects and Z3
;;gdscript          ; the language you waited for
;;(go +lsp)         ; the hipster dialect
(haskell +lsp)  ; a language that's lazier than I am
;;hy                ; readability of scheme w/ speed of python
;;idris             ; a language you can depend on
;;json              ; At least it ain't XML
;;(java +meghanada) ; the poster child for carpal tunnel syndrome
;;javascript        ; all(hope(abandon(ye(who(enter(here))))))
;;julia             ; a better, faster MATLAB
;;kotlin            ; a better, slicker Java(Script)
latex             ; writing papers in Emacs has never been so fun
;;lean
;;factor
;;ledger            ; an accounting system in Emacs
;;lua               ; one-based indices? one-based indices
markdown          ; writing docs for people to ignore
;;nim               ; python + lisp at the speed of c
;;nix               ; I hereby declare "nix geht mehr!"
;;ocaml             ; an objective camel
(org +noter)               ; organize your plain life in plain text
;;php               ; perl's insecure younger brother
;;plantuml          ; diagrams for confusing people more
;;purescript        ; javascript, but functional
(python +lsp)           ; beautiful is better than ugly
;;qt                ; the 'cutest' gui framework ever
;;racket            ; a DSL for DSLs
;;raku              ; the artist formerly known as perl6
;;rest              ; Emacs as a REST client
;;rst               ; ReST in peace
;;(ruby +rails)     ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
;;rust              ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
(scala +lsp)             ; java, but good
;;scheme            ; a fully conniving family of lisps
sh                ; she sells {ba,z,fi}sh shells on the C xor
;;sml
;;solidity          ; do you need a blockchain? No.
;;swift             ; who asked for emoji variables?
;;terra             ; Earth and Moon in alignment for performance.
;;web               ; the tubes
;;yaml              ; JSON, but readable
#+end_src

*** Other things
Email
#+name: doom-email
#+begin_src emacs-lisp
;;(mu4e +gmail)
;;notmuch
;;(wanderlust +gmail)
#+end_src

Apps
#+name: doom-app
#+begin_src emacs-lisp
;;calendar
;;everywhere        ; *leave* Emacs!? You must be joking
;;irc               ; how neckbeards socialize
;;(rss +org)        ; emacs as an RSS reader
;;twitter           ; twitter client https://twitter.com/vnought
#+end_src

Configuration
#+name: doom-config
#+begin_src emacs-lisp
literate
(default +bindings +smartparens)
#+end_src

** Packages
A literate packages.el. This file is used to declare external packages. Declarations are done beside packages configurations, [[*Beacon][like so]].
#+begin_src emacs-lisp :tangle "packages.el" :comments no
;; -*- no-byte-compile: t; -*-
#+end_src

* General
#+begin_src emacs-lisp
(setq default-directory "~"
      locale-coding-system 'utf-8-unix
      mouse-wheel-tilt-scroll 't)

(custom-set-faces! `(link :foreground ,(doom-color 'violet) :weight bold :underline t ))
#+end_src

** Widows and buffers
Keybindings for convenient buffer and window manipulation
#+begin_src emacs-lisp
(map! :leader
      :desc "Switch workspace buffer other window" :n "bw" #'+ivy/switch-workspace-buffer-other-window
      :desc "Switch buffer other window" :n "bW" #'+ivy/switch-buffer-other-window
      :desc "Kill buffer and window" :n "bD" #'kill-buffer-and-window
      :desc "Window nav hydra" :n "wn" #'+hydra/window-nav/body)
#+end_src

Split windows vertically by default
#+begin_src emacs-lisp
(setq split-height-threshold nil)
(setq split-width-threshold 0)
#+end_src

** Popups
Don't use popup system for info-mode.
#+begin_src emacs-lisp
(set-popup-rule! "^\\*info\\*$" :ignore t)
#+end_src

** Dictionary lookup
Functions to lookup random words in the dictionary definitions and synonyms
#+begin_src emacs-lisp
(defun ak/dictionary-definition-prompt (identifier &optional arg)
  "Look up the definition of any word."
  (interactive
   (list (read-string "Look up in dictionary: ")
         current-prefix-arg))
  (message "Looking up dictionary definition for %S" identifier)
  (cond ((and IS-MAC (require 'osx-dictionary nil t))
         (osx-dictionary--view-result identifier))
        ((and +lookup-dictionary-prefer-offline
              (require 'wordnut nil t))
         (unless (executable-find wordnut-cmd)
           (user-error "Couldn't find %S installed on your system"
                       wordnut-cmd))
         (wordnut-search identifier))
        ((require 'define-word nil t)
         (define-word identifier nil arg))
        ((user-error "No dictionary backend is available"))))

(map! :leader
      :desc "Dictionary (w/ prompt)" :n "sx" #'ak/dictionary-definition-prompt)
#+end_src

* UI
** Appearance
#+begin_src emacs-lisp
(setq doom-font (font-spec :family "Source Code Pro" :size 16 :weight 'Regular)
      doom-theme 'doom-gruvbox-mod
      display-line-numbers-type 'relative
      display-time-default-load-average nil)

(add-hook 'doom-after-init-modules-hook #'doom/reload-theme 100)
#+end_src

*** Initial screen
#+begin_src emacs-lisp
;;(toggle-frame-fullscreen)
(add-to-list 'default-frame-alist '(fullscreen . maximized))
#+end_src

*** Dashboard
Hide cursor and turn off line highlighting in dashboard
#+begin_src emacs-lisp
(setq-hook! '+doom-dashboard-mode-hook hl-line-mode -1)
(setq-hook! '+doom-dashboard-mode-hook evil-normal-state-cursor (list nil))
#+end_src

Hide the dashboard menu
#+begin_src emacs-lisp
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)
#+end_src

**** Splash image
Fancy splash image yanked from [[https://github.com/tecosaur/emacs-config/blob/master/config.org#splash-screen][here]]
#+begin_src emacs-lisp
(defvar fancy-splash-image-template
  (expand-file-name "emacs-e-template.svg" doom-private-dir)
  "Default template svg used for the splash image, with substitutions from ")

(defvar fancy-splash-sizes
  `((:height 300 :min-height 50 :padding (0 . 2))
    (:height 250 :min-height 42 :padding (2 . 4))
    (:height 200 :min-height 35 :padding (3 . 3))
    (:height 150 :min-height 28 :padding (3 . 3))
    (:height 100 :min-height 20 :padding (2 . 2))
    (:height 75  :min-height 15 :padding (2 . 1))
    (:height 50  :min-height 10 :padding (1 . 0))
    (:height 1   :min-height 0  :padding (0 . 0)))
  "list of plists with the following properties
  :height the height of the image
  :min-height minimum `frame-height' for image
  :padding `+doom-dashboard-banner-padding' (top . bottom) to apply
  :template non-default template file
  :file file to use instead of template")

(defvar fancy-splash-template-colours
  '(("$colour1" . keywords) ("$colour2" . type) ("$colour3" . base5) ("$colour4" . base8))
  "list of colour-replacement alists of the form (\"$placeholder\" . 'theme-colour) which applied the template")

(unless (file-exists-p (expand-file-name "theme-splashes" doom-cache-dir))
  (make-directory (expand-file-name "theme-splashes" doom-cache-dir) t))

(defun fancy-splash-filename (theme-name height)
  (expand-file-name (concat (file-name-as-directory "theme-splashes")
                            theme-name
                            "-" (number-to-string height) ".svg")
                    doom-cache-dir))

(defun fancy-splash-clear-cache ()
  "Delete all cached fancy splash images"
  (interactive)
  (delete-directory (expand-file-name "theme-splashes" doom-cache-dir) t)
  (message "Cache cleared!"))

(defun fancy-splash-generate-image (template height)
  "Read TEMPLATE and create an image if HEIGHT with colour substitutions as
   described by `fancy-splash-template-colours' for the current theme"
  (with-temp-buffer
    (insert-file-contents template)
    (re-search-forward "$height" nil t)
    (replace-match (number-to-string height) nil nil)
    (re-search-forward "$width" nil t)
    (replace-match (number-to-string height) nil nil)
    (dolist (substitution fancy-splash-template-colours)
      (goto-char (point-min))
      (while (re-search-forward (car substitution) nil t)
        (replace-match (doom-color (cdr substitution)) nil nil)))
    (write-region nil nil
                  (fancy-splash-filename (symbol-name doom-theme) height) nil nil)))

(defun fancy-splash-generate-images ()
  "Perform `fancy-splash-generate-image' in bulk"
  (dolist (size fancy-splash-sizes)
    (unless (plist-get size :file)
      (fancy-splash-generate-image (or (plist-get size :template)
                                       fancy-splash-image-template)
                                   (plist-get size :height)))))

(defun ensure-theme-splash-images-exist (&optional height)
  (unless (file-exists-p (fancy-splash-filename
                          (symbol-name doom-theme)
                          (or height
                              (plist-get (car fancy-splash-sizes) :height))))
    (fancy-splash-generate-images)))

(defun get-appropriate-splash ()
  (let ((height (frame-height)))
    (cl-some (lambda (size) (when (>= height (plist-get size :min-height)) size))
             fancy-splash-sizes)))

(setq fancy-splash-last-size nil)
(setq fancy-splash-last-theme nil)
(defun set-appropriate-splash (&rest _)
  (let ((appropriate-image (get-appropriate-splash)))
    (unless (and (equal appropriate-image fancy-splash-last-size)
                 (equal doom-theme fancy-splash-last-theme)))
    (unless (plist-get appropriate-image :file)
      (ensure-theme-splash-images-exist (plist-get appropriate-image :height)))
    (setq fancy-splash-image
          (or (plist-get appropriate-image :file)
              (fancy-splash-filename (symbol-name doom-theme) (plist-get appropriate-image :height))))
    (setq +doom-dashboard-banner-padding (plist-get appropriate-image :padding))
    (setq fancy-splash-last-size appropriate-image)
    (setq fancy-splash-last-theme doom-theme)
    (+doom-dashboard-reload)))

(add-hook 'window-size-change-functions #'set-appropriate-splash)
(add-hook 'doom-load-theme-hook #'set-appropriate-splash)
#+end_src

** Modeline
#+begin_src emacs-lisp
(setq doom-modeline-buffer-encoding nil
      doom-modeline-major-mode-icon t
      doom-modeline-modal-icon nil)
#+end_src

Key binding to toggle modeline
#+begin_src emacs-lisp
(map! :leader
      :desc "Modeline" :n "tm"
      #'hide-mode-line-mode)
#+end_src

*** Evil state indicators
Change faces of evil indicators
#+begin_src emacs-lisp
(custom-set-faces!
  `(doom-modeline-evil-emacs-state :background ,(doom-color 'violet) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-normal-state :background ,(doom-color 'green) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-insert-state :background ,(doom-color 'dark-blue) :foreground ,(doom-color 'fg))
  `(doom-modeline-evil-visual-state :background ,(doom-color 'orange) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-motion-state :background ,(doom-color 'red) :foreground ,(doom-color 'fg))
  `(doom-modeline-evil-operator-state :background ,(doom-color 'yellow) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-replace-state :background ,(doom-color 'magenta) :foreground ,(doom-color 'fg)))
#+end_src

*** Battery and time
Display battery and time in modeline
#+begin_src emacs-lisp
;; (display-battery-mode 1)
;; (display-time-mode 1)
#+end_src

*** TODO Change `doom-modeline-buffer-file-name' based on window size

** Beacon
Beacon to never lose my cursor again
#+begin_src emacs-lisp :tangle "packages.el" :comments no
(package! beacon)
#+end_src

#+begin_src emacs-lisp
(use-package! beacon
  :hook (doom-after-init-modules . beacon-mode)
  :config
  (defun ak/set-beacon-theme-color-h ()
    "Set the beacon theme color to orange according to current theme"
    (setq beacon-color (doom-color 'orange)))
  (add-hook! 'doom-load-theme-hook :append
            #'ak/set-beacon-theme-color-h))
#+end_src

* Tools
** Treemacs
#+begin_src emacs-lisp
(setq doom-themes-treemacs-theme "doom-colors")
#+end_src

** Dired
Change dired's default behaviour of creating new buffers and use vim like keybindings
#+begin_src emacs-lisp
(after! dired
  (map! :map dired-mode-map
        :desc "Go to parent directory" :n "h" (lambda () (interactive) (find-alternate-file ".."))
        :desc "Go to current entry" :n "l" #'dired-find-alternate-file
        :desc "Kill marked lines" :n "K" #'dired-do-kill-lines
        :desc "Goto home directory" :n "gh" (lambda () (interactive) (find-alternate-file (getenv "HOME"))))
  (setq dired-listing-switches "-agho --group-directories-first"))
#+end_src

*** Dired open
Dired open for easily opening files
#+begin_src emacs-lisp :tangle "packages.el" :comments no
(package! dired-open)
#+end_src

Configure dired-open for commonly opened file types
#+begin_src emacs-lisp
(use-package! dired-open
  :commands (dired)
  :config
  (setq dired-open-extensions '(("vcd" . "gtkwave"))))
#+end_src

** Projectile
Keep projectile from adding new projects everytime a directory is opened
#+begin_src emacs-lisp
(setq projectile-track-known-projects-automatically nil)
#+end_src

** mu4e
Add mu4e to load path
#+begin_src emacs-lisp
;; (add-to-list 'load-path "/usr/share/emacs/site-lisp/mu4e")
#+end_src

Configure mu4e
#+begin_src emacs-lisp
;; (set-email-account! "gmail"
;;   '((mu4e-sent-folder       . "/gmail/Sent Mail")
;;     (mu4e-drafts-folder     . "/gmail/Drafts")
;;     (mu4e-trash-folder      . "/gmail/Trash")
;;     (mu4e-refile-folder     . "/gmail/All Mail")
;;     (smtpmail-smtp-user     . "arunkumarmv1997@gmail.com")
;;     (user-mail-address      . "arunkumarmv1997@gmail.com")    ;; only needed for mu < 1.4
;;     (mu4e-compose-signature . "---\nArunkumar M V"))
;;   t)
#+end_src

** Evil mode
Pin evil-collection to resolve [[https://github.com/hlissner/doom-emacs/issues/5454][this]] issue
#+begin_src emacs-lisp :tangle "packages.el" :comments no
(package! evil-collection :pin "47bc8571a6105d7cf61b7fd0fce4a87f8c2c4725")
#+end_src

Evil vim state indicators
#+begin_src emacs-lisp
(after! evil
  (setq-default
   evil-emacs-state-tag          " E "
   evil-normal-state-tag         " N "
   evil-insert-state-tag         " I "
   evil-visual-char-tag          " V "
   evil-visual-line-tag          " VL "
   evil-visual-screen-line-tag   " VSL "
   evil-visual-block-tag         " VB "
   evil-motion-state-tag         " M "
   evil-operator-state-tag       " O "
   evil-replace-state-tag        " R "))
#+end_src
** Document utilities
*** PDF tools
Open in midnight dark mode
#+begin_src emacs-lisp
(add-hook 'pdf-tools-enabled-hook 'pdf-view-midnight-minor-mode)
(setq-default pdf-view-display-size 'fit-page)
#+end_src

Key bindings to scroll to horizontal extremes when zoomed in
#+begin_src emacs-lisp
(map! :map pdf-view-mode-map
      :desc "Scroll to extreme left of page" :n "C-c h" (lambda ()
                                                          (interactive)
                                                          (image-bol 1))
      :desc "Scroll to extreme right of page" :n "C-c l" (lambda ()
                                                           (interactive)
                                                           (image-eol 1)))
#+end_src

Key bindings to move forward and back in pdf history
#+begin_src emacs-lisp
(map! :map pdf-view-mode-map
      :desc "Forward" :n "ghf" #'pdf-history-forward
      :desc "Backward" :n "ghb" #'pdf-history-backward)
#+end_src

*** Nov
Nov.el for reading EPUBs
#+begin_src emacs-lisp :tangle "packages.el" :comments no
(package! nov)
#+end_src

Configure nov-mode. Open epubs with nov, set font and column width.
#+begin_src emacs-lisp
(use-package! nov
  :config
  (add-to-list 'auto-mode-alist '("\\.epub\\'" . nov-mode))

  (defun ak/nov-font-setup ()
    (face-remap-add-relative 'variable-pitch :family "Bookerly"
                             :height 1.0))

  (add-hook 'nov-mode-hook 'ak/nov-font-setup)
  (setq nov-text-width 80))
#+end_src
*** Markdown
**** Polymode markdown
For markdown syntax highlighting
#+begin_src emacs-lisp :tangle "packages.el" :comments no
(package! poly-markdown :pin "1536cf0c32f71d5cd05c90f7905905e38006e95d")
#+end_src

* Org mode
#+begin_src emacs-lisp
(after! org
  (setq org-startup-folded 'content)
  (add-to-list 'org-file-apps '("\\.pdf\\'" . emacs)))
#+end_src

Right align org tags
#+begin_src emacs-lisp
(setq org-tags-column 60)
#+end_src

** Directories and files
#+begin_src emacs-lisp
(setq org-directory (concat (getenv "HOME") "/Org")

doom/config.el
(setq default-directory "~")

(setq
 doom-font (font-spec :family "Source Code Pro for Powerline" :size 17 :weight 'Regular)
 doom-theme 'doom-gruvbox-mod
 display-line-numbers-type 'relative
 display-time-default-load-average nil)

(display-battery-mode 1)
(display-time-mode 1)

;;(add-to-list 'initial-frame-alist '(fullscreen . maximized))

(toggle-frame-fullscreen)

(when (featurep! :ui treemacs)
  (remove-hook 'doom-load-theme-hook #'doom-themes-treemacs-config))

(map!
 :map dired-mode-map
 :n "-" (lambda () (interactive) (find-alternate-file "..")))

(setq projectile-track-known-projects-automatically nil)

(setq-hook! '+doom-dashboard-mode-hook hl-line-mode -1)
(setq-hook! '+doom-dashboard-mode-hook evil-normal-state-cursor (list nil))

(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)

(defvar fancy-splash-image-template
  (expand-file-name "emacs-e-template.svg" doom-private-dir)
  "Default template svg used for the splash image, with substitutions from ")

(defvar fancy-splash-sizes
  `((:height 300 :min-height 50 :padding (0 . 2))
    (:height 250 :min-height 42 :padding (2 . 4))
    (:height 200 :min-height 35 :padding (3 . 3))
    (:height 150 :min-height 28 :padding (3 . 3))
    (:height 100 :min-height 20 :padding (2 . 2))
    (:height 75  :min-height 15 :padding (2 . 1))
    (:height 50  :min-height 10 :padding (1 . 0))
    (:height 1   :min-height 0  :padding (0 . 0)))
  "list of plists with the following properties
  :height the height of the image
  :min-height minimum `frame-height' for image
  :padding `+doom-dashboard-banner-padding' (top . bottom) to apply
  :template non-default template file
  :file file to use instead of template")

(defvar fancy-splash-template-colours
  '(("$colour1" . keywords) ("$colour2" . type) ("$colour3" . base5) ("$colour4" . base8))
  "list of colour-replacement alists of the form (\"$placeholder\" . 'theme-colour) which applied the template")

(unless (file-exists-p (expand-file-name "theme-splashes" doom-cache-dir))
  (make-directory (expand-file-name "theme-splashes" doom-cache-dir) t))

(defun fancy-splash-filename (theme-name height)
  (expand-file-name (concat (file-name-as-directory "theme-splashes")
                            theme-name
                            "-" (number-to-string height) ".svg")
                    doom-cache-dir))

(defun fancy-splash-clear-cache ()
  "Delete all cached fancy splash images"
  (interactive)
  (delete-directory (expand-file-name "theme-splashes" doom-cache-dir) t)
  (message "Cache cleared!"))

(defun fancy-splash-generate-image (template height)
  "Read TEMPLATE and create an image if HEIGHT with colour substitutions as
   described by `fancy-splash-template-colours' for the current theme"
  (with-temp-buffer
    (insert-file-contents template)
    (re-search-forward "$height" nil t)
    (replace-match (number-to-string height) nil nil)
    (re-search-forward "$width" nil t)
    (replace-match (number-to-string height) nil nil)
    (dolist (substitution fancy-splash-template-colours)
      (goto-char (point-min))
      (while (re-search-forward (car substitution) nil t)
        (replace-match (doom-color (cdr substitution)) nil nil)))
    (write-region nil nil
                  (fancy-splash-filename (symbol-name doom-theme) height) nil nil)))

(defun fancy-splash-generate-images ()
  "Perform `fancy-splash-generate-image' in bulk"
  (dolist (size fancy-splash-sizes)
    (unless (plist-get size :file)
      (fancy-splash-generate-image (or (plist-get size :template)
                                       fancy-splash-image-template)
                                   (plist-get size :height)))))

(defun ensure-theme-splash-images-exist (&optional height)
  (unless (file-exists-p (fancy-splash-filename
                          (symbol-name doom-theme)
                          (or height
                              (plist-get (car fancy-splash-sizes) :height))))
    (fancy-splash-generate-images)))

(defun get-appropriate-splash ()
  (let ((height (frame-height)))
    (cl-some (lambda (size) (when (>= height (plist-get size :min-height)) size))
             fancy-splash-sizes)))

(setq fancy-splash-last-size nil)
(setq fancy-splash-last-theme nil)
(defun set-appropriate-splash (&rest _)
  (let ((appropriate-image (get-appropriate-splash)))
    (unless (and (equal appropriate-image fancy-splash-last-size)
                 (equal doom-theme fancy-splash-last-theme)))
    (unless (plist-get appropriate-image :file)
      (ensure-theme-splash-images-exist (plist-get appropriate-image :height)))
    (setq fancy-splash-image
          (or (plist-get appropriate-image :file)
              (fancy-splash-filename (symbol-name doom-theme) (plist-get appropriate-image :height))))
    (setq +doom-dashboard-banner-padding (plist-get appropriate-image :padding))
    (setq fancy-splash-last-size appropriate-image)
    (setq fancy-splash-last-theme doom-theme)
    (+doom-dashboard-reload)))

(add-hook 'window-size-change-functions #'set-appropriate-splash)
(add-hook 'doom-load-theme-hook #'set-appropriate-splash)

(setq
 doom-modeline-buffer-encoding nil
 doom-modeline-major-mode-icon t
 doom-modeline-modal-icon nil)

(custom-set-faces!
  `(doom-modeline-evil-emacs-state :background ,(doom-color 'violet) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-normal-state :background ,(doom-color 'green) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-insert-state :background ,(doom-color 'dark-blue) :foreground ,(doom-color 'fg))
  `(doom-modeline-evil-visual-state :background ,(doom-color 'orange) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-motion-state :background ,(doom-color 'red) :foreground ,(doom-color 'fg))
  `(doom-modeline-evil-operator-state :background ,(doom-color 'yellow) :foreground ,(doom-color 'base0))
  `(doom-modeline-evil-replace-state :background ,(doom-color 'magenta) :foreground ,(doom-color 'fg)))

(map! :leader
      :desc "Modeline" :n "tm"
      #'doom-modeline-mode)

(setq-default
 evil-emacs-state-tag          " E "
 evil-normal-state-tag         " N "
 evil-insert-state-tag         " I "
 evil-visual-char-tag          " V "
 evil-visual-line-tag          " VL "
 evil-visual-screen-line-tag   " VSL "
 evil-visual-block-tag         " VB "
 evil-motion-state-tag         " M "
 evil-operator-state-tag       " O "
 evil-replace-state-tag        " R ")

(add-hook 'lsp-ui-doc-mode-hook
            (lambda ()
              (when lsp-ui-doc-mode
                (remove-hook 'post-command-hook #'lsp-ui-doc--make-request t))))

(map!
 :map lsp-ui-doc-mode-map
 :n "gh" #'lsp-ui-doc-glance)

(setq lsp-log-io nil
      lsp-python-ms-extra-paths ["./src/python" "./configs"])

(after! python
  (setq conda-env-home-directory (expand-file-name "~/.conda"))
  (custom-set-variables
   '(conda-anaconda-home (getenv "CONDA_HOME"))))

(after! lsp-clients
  (set-lsp-priority! 'clangd 1))  ; ccls has priority 0

(setq lsp-clients-clangd-args '("-j=3"
                                "--background-index"
                                "--clang-tidy"
                                "--completion-style=detailed"
                                "--header-insertion=never"))

(setq org-directory "~/Org/"
      org-startup-folded 'content)

(after! org
  (setq org-todo-keywords '((sequence "TODO(t)" "INPROGRESS(i)" "PARKED(p)" "|" "DONE(d)" "KILLED(k)")))

  (defface +org-todo-todo '((t)) "Face for org todo keyword TODO")
  (defface +org-todo-inprogress '((t)) "Face for org todo keyword INPROGRESS")
  (defface +org-todo-parked '((t)) "Face for org todo keyword PARKED")
  (defface +org-todo-done '((t)) "Face for org todo keyword DONE")
  (defface +org-todo-killed '((t)) "Face for org todo keyword KILLED")

  (custom-set-faces!
    `(+org-todo-todo :foreground ,(doom-color 'yellow) :inherit (bold-italic org-todo))
    `(+org-todo-inprogress :foreground ,(doom-color 'violet) :inherit (bold-italic org-todo))
    `(+org-todo-parked :foreground ,(doom-color 'orange) :inherit (bold-italic org-todo))
    `(+org-todo-done :foreground ,(doom-color 'green) :inherit (bold-italic org-todo))
    `(+org-todo-killed :foreground ,(doom-color 'red) :inherit (bold-italic org-todo)))

  (setq org-todo-keyword-faces
        '(("TODO" . +org-todo-todo)
          ("INPROGRESS" . +org-todo-inprogress)
          ("PARKED" . +org-todo-parked)
          ("DONE" . +org-todo-done)
          ("KILLED" . +org-todo-killed))))

(use-package! org-fancy-priorities
  :hook (org-mode . org-fancy-priorities-mode)
  :hook (org-agenda-mode . org-fancy-priorities-mode)
  :config (setq org-fancy-priorities-list '("■" "■" "■")))

(after! org
  (setq org-tags-column 60))

(add-hook! 'org-mode-hook 'org-zotxt-mode)
(after! org
  (add-to-list 'org-file-apps '("\\.pdf\\'" . emacs)))

(defun org-zotxt-insert-current-selection ()
  "Insert reference link for the currently selected item in Zotero"
  (interactive)
  (org-zotxt-insert-reference-link 4))

(map! :map org-mode-map
      :localleader
      (:prefix ("z" . "zotero")
       :desc "Link to selected item" "i" #'org-zotxt-insert-current-selection
       :desc "Link to an item"       "I" #'org-zotxt-insert-reference-link
       :desc "Open link"             "a" #'org-zotxt-open-attachment))

(after! (org zotxt)
  (org-link-set-parameters "zotero"
                           :face `(:foreground ,(doom-color 'red)
                                   :weight bold
                                   :slant italic)))

(add-hook! 'org-mode-hook 'org-beamer-mode)
(require 'ox-latex)

(setq org-latex-listings t)
(add-to-list 'org-latex-packages-alist '("" "listings"))
(add-to-list 'org-latex-packages-alist '("" "color"))

(setq vterm-shell "/bin/zsh")

(map!
 :map vterm-mode-map
 :n "-" #'vterm-send-up
 :n "=" #'vterm-send-down)

(map! :leader
      :desc "Edit .env" :n "fv"
      (lambda ()
          (interactive)
          (find-file-other-window (expand-file-name "~/.env")))
      :desc "Edit .zshrc" :n "fz"
      (lambda ()
          (interactive)
          (find-file-other-window (expand-file-name "../zsh/.zshrc" doom-private-dir))))
